crates/protocol/src/deposits.rs

```diff
@@ -0,0 +1,688 @@
+//! Contains deposit transaction types and helper methods.
+#[cfg(not(feature = "std"))]
+use alloc::{string::String, vec::Vec};
+use alloy_eips::eip2718::Encodable2718;
+use alloy_primitives::{b256, keccak256, Address, Bytes, Log, TxKind, B256, U256, U64};
+use alloy_rlp::Encodable;
+use core::fmt::Display;
+use op_alloy_consensus::{OpTxEnvelope, TxDeposit};
+
+/// Deposit log event abi signature.
+pub const DEPOSIT_EVENT_ABI: &str = "TransactionDeposited(address,address,uint256,bytes)";
+
+/// Deposit event abi hash.
+///
+/// This is the keccak256 hash of the deposit event ABI signature.
+/// `keccak256("TransactionDeposited(address,address,uint256,bytes)")`
+pub const DEPOSIT_EVENT_ABI_HASH: B256 =
+    b256!("b3813568d9991fc951961fcb4c784893574240a28925604d09fc577c55bb7c32");
+
+/// The initial version of the deposit event log.
+pub const DEPOSIT_EVENT_VERSION_0: B256 = B256::ZERO;
+
+/// An [op_alloy_consensus::TxDeposit] validation error.
+#[derive(Debug)]
+pub enum DepositError {
+    /// Unexpected number of deposit event log topics.
+    UnexpectedTopicsLen(usize),
+    /// Invalid deposit event selector.
+    /// Expected: [B256] (deposit event selector), Actual: [B256] (event log topic).
+    InvalidSelector(B256, B256),
+    /// Incomplete opaqueData slice header (incomplete length).
+    IncompleteOpaqueData(usize),
+    /// The log data is not aligned to 32 bytes.
+    UnalignedData(usize),
+    /// Failed to decode the `from` field of the deposit event (the second topic).
+    FromDecode(B256),
+    /// Failed to decode the `to` field of the deposit event (the third topic).
+    ToDecode(B256),
+    /// Invalid opaque data content offset.
+    InvalidOpaqueDataOffset(Bytes),
+    /// Invalid opaque data content length.
+    InvalidOpaqueDataLength(Bytes),
+    /// Opaque data length exceeds the deposit log event data length.
+    /// Specified: [usize] (data length), Actual: [usize] (opaque data length).
+    OpaqueDataOverflow(usize, usize),
+    /// Opaque data with padding exceeds the specified data length.
+    PaddedOpaqueDataOverflow(usize, usize),
+    /// An invalid deposit version.
+    InvalidVersion(B256),
+    /// Unexpected opaque data length
+    UnexpectedOpaqueDataLen(usize),
+    /// Failed to decode the deposit mint value.
+    MintDecode(Bytes),
+    /// Failed to decode the deposit gas value.
+    GasDecode(Bytes),
+    /// A custom error wrapping [anyhow::Error].
+    Custom(anyhow::Error),
```

Comment:
I think we should use a generic here that implements the [`core::error::Error`](https://doc.rust-lang.org/core/error/trait.Error.html) trait.

e.g.
```rust
pub enum DepositError<T: impl core::error::Error> {
	...
	/// A custom [core::error::Error].
	Custom(T)
}
```